### DISCUSSION TITLE
Code extension for DataRef name factorization

### DISCUSSION THREAD

**Discussion by GioCC** [curiosity-workshop/XPlane-Direct-superceded#11] [Ideas]
Hi, 
while having a look at the code, i was thinking about a possible improvement / extension.

In a more complex scenario, there are quite a few dataref strings that have a common prefix, which accumulate to a not negligible total length.
Taking as example the MCP sketch, most datarefs have a `laminar/B738/autopilot/` prefix in common; there are about 50 datarefs times 23 characters, amounting to 1200 chars total.  
I was thinking about having an additional version of each `registerDataRef()` and `registerCommand()` function that take a double string parameter, e.g.:
`int registerDataRef(char* name, int, unsigned int, char*)` ->
`int registerDataRef(char* prefix, char* name, int, unsigned int, char*)`

This could obviously apply to both the RAM and PROGMEM versions; examples below refer to the latter.

Example of invocation: from e.g.
```
  Xinterface.registerDataRef(F("laminar/B738/autopilot/n1_status"), XPL_READ, 10, 0, &n1_status);
  Xinterface.registerDataRef(F("laminar/B738/autopilot/speed_status1"), XPL_READ, 10, 0, &speed_status1);
```
to:
```
  const __FlashStringHelper* prefix = F("laminar/B738/autopilot/");
  Xinterface.registerDataRef(prefix, F("n1_status"),     XPL_READ, 10, 0, &n1_status);
  Xinterface.registerDataRef(prefix, F("speed_status1"), XPL_READ, 10, 0, &speed_status1);
```
Beside the additional functions mentioned (but the current ones might become as short as simple redirections to the "extended" ones), there would be following impacts:
- two additional pointers would be required in the `_dataRefStructure` and `_commandStructure` structs -> penalty of 4 bytes for each dataref;
- the dataref request message composition would bear almost no additional cost (basically one more `%S` in the `sprintf()` format string));
- the processing of the reply message would entail recomposing the complete string in a temporary buffer for the comparison, once for each dataref.

Do you think this extension would make sense or be useful anyway?
If you're interested I could have a go at implementing those modifications and see what the results are .

Looking forward to your opinion!
Giorgio Croci Candiani

**Comment by curiosity-workshop** [comment]
I’ve been considering tokenizing all of the useful data refs, so the
strings are stored on the host and identified as type int for the token.
Would that make more sense?

#define laminar_b738_autopilot_somebutton 1

RegisterDataRef(laminar_b738_autopilot_somebutton, ...

Something like that.

**Reply by GioCC** [reply]
Thanks for the quick reply!

My opinion as a developer: at a first glance it looks attractive and definitely efficient.

My opinion from a user's standpoint: maybe it's less intimidating to write a dataref name directly, rather than to have to look for the right code/name in yet another loooong list of entries (even if I understand that the dataref name and the constant's name would be practically identical). 

But most of all, I'm afraid it would create a few ugly constraints by not allowing to address custom or newly introduced datarefs if the plugin isn't correspondingly updated.

**Comment by curiosity-workshop** [comment]
I was considering it as an expansion rather than a replacement, so that the
new token method could be used for applications using a lot of data refs.
I’m also looking at a plug-in based solution as well, sorta like simvim but
not a closed sandbox.  Thank you so much for your ideas!

**Reply by GioCC** [reply]
You're quite right... as an expansion, it's effectively the best of both worlds, as long as it doesn't put too much burden on the plugin I guess. If you should decide to go that way I would like that solution.
Maybe if I have a little time in the coming days I might just go ahead and try an implementation of my idea, if something comes out of it I'll send you the code just in case it comes useful!

**Comment by curiosity-workshop** [comment]
I look forward to seeing it.  If it works well we can add it to the release
version.

**Reply by GioCC** [reply]
Hi,
I actually had some time in the weekend, so... I just did it.
Do you rather have me posting it here or should I send it to you as a private mail?

**Comment by curiosity-workshop** [comment]
You can post here

**Comment by GioCC** [comment]
Hi,
there you go.

For full disclosure, I did not have time to perform any test yet (except correct compilation obviously); I thought I'd send you the code anyway so you can have a look at what it's about.

A little summary of main modifications:
- I concentrated all 'register' functions to a single, private helper function (one for Datarefs and one for commands); the public declarations are now just used to establish the call signature and redirect to the helpers as inlines. This slightly reduces the program memory footprint and avoids code duplication (beside making code more manageable despite the increased number of functions);
- I replaced the dual string pointer for names (RAM/progmem version) with a union, compensating the space for the added prefix string, and introduced a flag to mark which version must be used;
- the comparison upon receiving an XPLRESPONSE_DATAREF message does not use any RAM buffer, but is split (if required) into two half-comparisons for prefix and name
- on XPLCMD_SENDREQUEST, an additional placeholder is used in the sprintf() format string, which is filled with an empty string if no prefix is used.

Despite the added features, compiling the very same sketch (basically the MCP example, just slightly modified) results in a negligible increase in occupied memory, both RAM and flash (1963/23168 bytes vs 1941/22780, <0.3%). I must confess that I had expected even a small decrease in space occupation, given the number of functions I removed. 
I also haven't yet tried adapting the sketch by exploiting the prefix feature, which should result in the actual space savings.

And BTW... the GCC mentions are just my initials, no relation to the compiler :)

Looking forward to your comments or suggestions.

[XPLDirect_GCC-alpha.zip](https://github.com/curiosity-workshop/XPlane-Direct/files/6295150/XPLDirect_GCC-alpha.zip)

**Comment by curiosity-workshop** [comment]
Nice.  Just glancing at it I am concerned that the progmem functions are
only available on certain boards, it may not compile on an M0 or other.  I
will spend more time with it when I get a chance.

**Reply by GioCC** [reply]
By all means, take your time. However, the code still has the compilation switches to exclude progmem stuff where it's not enabled, the same way as it was in the original code (unless of course I overlooked something), and there's non-progmem code for all the same features.
